kernel/kernel-api/src/main/java/io/delta/kernel/coordinatedcommits/CommitCoordinatorClient.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.coordinatedcommits;
+
+import io.delta.kernel.TableIdentifier;
+import io.delta.kernel.annotation.Evolving;
+import io.delta.kernel.annotation.Nullable;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.engine.coordinatedcommits.CommitFailedException;
+import io.delta.kernel.engine.coordinatedcommits.CommitResponse;
+import io.delta.kernel.engine.coordinatedcommits.GetCommitsResponse;
+import io.delta.kernel.engine.coordinatedcommits.UpdatedActions;
+import io.delta.kernel.engine.coordinatedcommits.actions.AbstractMetadata;
+import io.delta.kernel.engine.coordinatedcommits.actions.AbstractProtocol;
+import io.delta.kernel.utils.CloseableIterator;
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * The CommitCoordinatorClient is responsible for communicating with the commit coordinator and
+ * backfilling commits. It has four main APIs that need to be implemented:
+ *
+ * <ul>
+ *   <li>{@link #registerTable}: Determine the table config during commit coordinator registration.
+ *   <li>{@link #commit}: Commit a new version of the table.
+ *   <li>{@link #getCommits}: Tracks and returns unbackfilled commits.
+ *   <li>{@link #backfillToVersion}: Ensure that commits are backfilled if/when needed.
+ * </ul>
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public interface CommitCoordinatorClient {
+
+  /**
+   * Register the table represented by the given {@code logPath} at the provided {@code
+   * currentVersion} with the commit coordinator this commit coordinator client represents.
+   *
+   * <p>This API is called when the table is being converted from an existing file system table to a
+   * coordinated-commit table.
+   *
+   * <p>When a new coordinated-commit table is being created, the {@code currentVersion} will be -1
+   * and the upgrade commit needs to be a file system commit which will write the backfilled file
+   * directly.
```

Comment:
For my understanding you (the client, ie Spark/Kernel etc) call this first for some version N. Then when `commit` is called with version N, the CCC recognizes that this is the same version and thus the commit needs to be immediately backfilled/written to the filesystem?

Reply:
cc @LukasRupprecht @dhruvarya-db @prakharjain09

Reply:
Commit N would add the CC configuration to the table so it'll be available in version N+1. It is not in version N so the commit does not go through the newly added commit coordinator client but rather just through the file system, i.e. backfilling is not necessary.